src/devices/blueline.c

```diff
@@ -0,0 +1,330 @@
+ * - Second is a 2 byte payload, which is interpreted differently based on the two lowest bits of the first byte
+ * - Finally is a 1 byte CRC, calculated across the 2 payload bytes (not the header)
+ *
+ * The CRC is a CRC-8-ATM with polynomial 100000111, but it may be required to modify the payload bytes before
+ * calculating it depending on message type.
+ *
+ * There are 4 message types that can be indicated by the 2 lowest bits of the first payload byte:
+ * 0 - ID message (payload is not offset)
+ * 1 - power message (payload is offset)
+ * 2 - temperature/status message (payload is offset)
+ * 3 - energy message (payload is offset)
+ *
+ * For the ID message (0), the CRC can be calculated directly on the payload as sent, and when the payload is
+ * interpreted as a 16-bit integer it gives the ID of the transmitter.  This message is sent when the
+ * monitor is first powered on and if the button on the monitor is pressed briefly.  If the button on the
+ * monitor is held for >10 seconds, the monitor will change it's ID and report the new one.
+ *
+ * While the transmitter ID's are 16-bit, none of them can have any of the two lowest bits set or they
+ * would not be able to transmit their ID as message type 0, and when the offset is calculated (see below) they
+ * would also change the message type.
+ *
+ * For the 3 other message types, the payload must be offset before calculating the CRC or
+ * interpreting the data.  The offset is done by treating the whole payload as a single 16-bit integer and
+ * then subtracting the ID of the transmitter.  After the offset is done, then the CRC may be calculated
+ * and the payload may be interpreted.
+ *
+ * Note that if the transmitter's ID isn't known, the code can't easily determine if messages other than an
+ * ID payload are good or bad, and can't intepret their data correctly.  However, if the BLUELINE_USE_ID_TRANSMISSION
+ * or BLUELINE_USE_ID_AUTODETECT features are enabled, the system can learn the transmitter's ID by various methods.
+ *
+ * For the power message (1), the offset payload gives the number of milliseconds gap between impulses for the most
+ * recent impulses seen by the monitor.  To convert from this 'gap' to kilowatts, you will need your meter's
+ * Kh value. The Kh value is written obviously on the front of most meters, and 1.0 and 7.2 are very common.
+ *
+ * kW = (3600/gap) * Kh
+ *
+ * For the temperature message (2), the offset payload gives the temperature in an odd scaling in the last byte,
+ * and has some flag bits in the first byte.  The only known flag bit is the battery.
+ *
+ * For the energy message (3), the offset payload contains a continuously running power impulse accumulator. I'm
+ * not sure if there is a way to reset the accumulator.  The intended way to use it is to remember the accumulator
+ * value at the beginning of a time period, and then subtract that from the value at the end of the time period.
+ *
+ * kWh = 0.001 * (accumulated pulses) * Kh
```

Comment:
kh = kilo hour?

Reply:
That's just the standard name of the constant, and how it is labeled on most meters: https://ctlsys.com/support/kh_watthour_constant/